VirtoCommerce.CustomerModule.Data/Handlers/LogChangesMemberChangedEventHandler.cs

```diff
@@ -0,0 +1,249 @@
+                diff.Join(ObservedProperties, x => x.Name.ToLowerInvariant(), x => x.ToLowerInvariant(), (x, y) => x)
+                    .ToArray();
+            foreach (var difference in observedDifferences.Distinct(new DifferenceComparer()))
+            {
+                AddMemberPropertyChangedOperation(result, changedEntry, difference);
+            }
+
+            return result.Select(x => GetLogRecord(changedEntry.NewEntry, x));
+        }
+
+        protected virtual IEnumerable<string> GetContactChanges(Contact originalContact, Contact modifiedContact)
+        {
+            return
+                GetAddressChanges(
+                    originalContact, originalContact.Addresses, modifiedContact.Addresses)
+                    .Union(GetEmailChanges(originalContact, originalContact.Emails, modifiedContact.Emails))
+                    .Union(GetPhoneChanges(originalContact, originalContact.Phones, modifiedContact.Phones));
+        }
+
+        protected virtual IEnumerable<string> GetPhoneChanges(
+            Member member, IEnumerable<string> originalPhones, IEnumerable<string> modifiedPhones)
+        {
+            return GetListChanges(
+                originalPhones,
+                modifiedPhones,
+                result =>
+                    (state, source, target) =>
+                    {
+                        if (state == EntryState.Added)
+                        {
+                            AddPhoneAddedOperation(result, member, target);
+                        }
+                        else if (state == EntryState.Deleted)
+                        {
+                            AddPhoneRemovedOperation(result, member, target);
+                        }
+                    });
+        }
+
+        protected virtual IEnumerable<string> GetEmailChanges(
+            Member member, IEnumerable<string> originalEmails, IEnumerable<string> modifiedEmails)
+        {
+            return GetListChanges(
+                originalEmails,
+                modifiedEmails,
+                result =>
+                    (state, source, target) =>
+                    {
+                        if (state == EntryState.Added)
+                        {
+                            AddEmailAddedOperation(result, member, target);
+                        }
+                        else if (state == EntryState.Deleted)
+                        {
+                            AddEmailRemovedOperation(result, member, target);
+                        }
+                    });
+        }
+
+        protected virtual IEnumerable<string> GetAddressChanges(
+            Member member, IEnumerable<Address> originalAddress, IEnumerable<Address> modifiedAddress)
+        {
+            return GetListChanges(
+                originalAddress,
+                modifiedAddress,
+                result =>
+                    (state, source, target) =>
+                    {
+                        if (state == EntryState.Added)
```

Comment:
Why address changes aren't handled?
This if/else statements can be simplified by ` result.Add(
                string.Format($"ResourceManager.GetString(Address{state}", member.MemberType, member.Name, StringifyAddress(address)));` What do you think?

Reply:
Then it was a misunderstanding, I said that the Address doesn't have its own change history log (widget)

Reply:
Address changes are not handled because the changed address is considered as a new one, since `EqualityComparer` differentiates them. So we just get 2 logs instead of 1: `AddressDeleted` and `AddressAdded`. We could write a custom `EqualityComparer`, but does it make sense?